File: 브루트포스/프로그래머스Lv0-등수 매기기.py
"""
내 위로 몇명이 있는지? 를 찾으면 됨.
"""
from collections import defaultdict
def solution(score):
    table = defaultdict(int)
    for a, b in score:
        table[(a + b) / 2] += 1
    
    ranks = defaultdict(int)
    acc = 0
    for _score in sorted(table.keys(), reverse = True):
        if acc == 0:
            ranks[_score] = 1
        else:
            ranks[_score] = acc + 1
        acc += table[_score]
    
    result = []
    for a, b, in score:
        result.append(ranks[(a + b) / 2])
    
    return result
"""
정확성  테스트
테스트 1 〉	통과 (0.01ms, 10.3MB)
테스트 2 〉	통과 (0.02ms, 10.2MB)
테스트 3 〉	통과 (0.02ms, 9.96MB)
테스트 4 〉	통과 (0.01ms, 10.1MB)
테스트 5 〉	통과 (0.01ms, 10.1MB)
테스트 6 〉	통과 (0.01ms, 10.2MB)
테스트 7 〉	통과 (0.02ms, 10MB)
테스트 8 〉	통과 (0.01ms, 10.1MB)
테스트 9 〉	통과 (0.02ms, 10.2MB)
테스트 10 〉	통과 (0.02ms, 10.1MB)
테스트 11 〉	통과 (0.01ms, 10MB)
테스트 12 〉	통과 (0.01ms, 10.2MB)
"""

File: 브루트포스/프로그래머스Lv0-등수 매기기.js
// js에서는 딕셔너리로 변환하는 과정에서 Key가 문자열이되고, 
// 소수점까지 잘 처리가 안될 수 있어서 2로 나누는 평균이 아닌 그냥 안나누고 합으로 했음.
function solution(score) {
  const table = {}
  score.forEach(([a, b]) => {
      const key = a + b
      if (!table[key]) table[key] = 0
      table[key]++
  })
  
  const ranks = {}
  let acc = 0
  Object.keys(table)
      .sort((a, b) => parseInt(b, 10) - parseInt(a, 10))
      .forEach(_score => {
      if (acc === 0) ranks[_score] = 1
      else ranks[_score] = acc + 1
      acc += table[_score]
  })
  
  return score.map(([a, b]) => {
      const key = a + b
      return ranks[key]
  })
}
// 정확성  테스트
// 테스트 1 〉	통과 (0.12ms, 33.4MB)
// 테스트 2 〉	통과 (0.34ms, 33.5MB)
// 테스트 3 〉	통과 (0.14ms, 33.6MB)
// 테스트 4 〉	통과 (0.15ms, 33.4MB)
// 테스트 5 〉	통과 (0.18ms, 33.5MB)
// 테스트 6 〉	통과 (0.12ms, 33.4MB)
// 테스트 7 〉	통과 (0.13ms, 33.4MB)
// 테스트 8 〉	통과 (0.15ms, 33.5MB)
// 테스트 9 〉	통과 (0.16ms, 33.4MB)
// 테스트 10 〉	통과 (0.13ms, 33.6MB)
// 테스트 11 〉	통과 (0.20ms, 33.5MB)
// 테스트 12 〉	통과 (0.13ms, 33.6MB)